File: 04_Prompt_Engineering_and_Structured_Output/Exercises/python/02_validation_retry_loop.py
"""
Exercise 2 — Validation, Retry, and Feedback Loops (Task 4.4)
Domain: Prompt Engineering & Structured Output

Covers:
  - A semantic validation check that tool_use / JSON-schema compliance
    CANNOT catch on its own (line items not summing to the stated total)
  - Retry-with-error-feedback: appending the SPECIFIC validation error
    to the next request so the model can self-correct
  - Recognizing the case where retrying will NOT help, because the
    required information is genuinely absent from the source document
    (not a format problem — an absence problem)
"""

import anthropic

client = anthropic.Anthropic()

EXTRACT_INVOICE_TOOL = {
    "name": "extract_invoice",
    "description": "Record structured fields extracted from an invoice.",
    "input_schema": {
        "type": "object",
        "properties": {
            "vendor_name": {"type": "string"},
            "line_items": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "description": {"type": "string"},
                        "amount": {"type": "number"}
                    },
                    "required": ["description", "amount"]
                }
            },
            "stated_total": {"type": "number"},
            "purchase_order_number": {
                "type": ["string", "null"],
                "description": "PO number, or null if not present in the document."
            }
        },
        "required": ["vendor_name", "line_items", "stated_total"]
    }
}

# This invoice is deliberately malformed in a way a model can plausibly
# mis-extract on the first pass (the "Rush fee" line is easy to miss
# because it's on its own line without a clear "Item:" label) — a
# FORMAT/STRUCTURAL issue, which retry-with-feedback CAN fix.
INVOICE_WITH_FIXABLE_ISSUE = """
INVOICE #4471 — Sunrise Logistics LLC
  Freight handling ................. $310.00
  Fuel surcharge .................... $45.00
  Rush fee
    (applied for same-day dispatch)... $60.00
TOTAL DUE: $415.00
"""

# This invoice genuinely never states a PO number anywhere. No amount
# of retrying will produce one — the correct behavior is to accept
# purchase_order_number = null, not keep re-asking.
INVOICE_WITHOUT_PO_NUMBER = """
INVOICE — Cedar Grove Office Supply
  Paper (10 reams) .................. $65.00
  Toner cartridges (x2) ............. $140.00
TOTAL DUE: $205.00
"""


def validate_semantics(extraction: dict) -> str | None:
    """Returns a specific error string, or None if the extraction is valid.

    This is exactly the layer tool_use/JSON-schema enforcement does NOT
    provide: the schema is satisfied (right types, right shape) but the
    values may still be semantically wrong.
    """
    computed = sum(item["amount"] for item in extraction["line_items"])
    stated = extraction["stated_total"]
    if abs(computed - stated) > 0.01:
        return (
            f"Line items sum to {computed:.2f}, but stated_total is {stated:.2f}. "
            f"Re-read the document for a line item you may have missed or misread."
        )
    return None


def extract_with_retry(document_text: str, max_retries: int = 2) -> dict:
    messages = [{
        "role": "user",
        "content": f"Extract the invoice fields from this document:\n\n{document_text}"
    }]

    for attempt in range(max_retries + 1):
        response = client.messages.create(
            model="claude-sonnet-5",
            max_tokens=1024,
            tools=[EXTRACT_INVOICE_TOOL],
            tool_choice={"type": "tool", "name": "extract_invoice"},
            messages=messages,
        )
        tool_block = next(b for b in response.content if b.type == "tool_use")
        result = tool_block.input

        error = validate_semantics(result)
        if error is None:
            print(f"  [attempt {attempt + 1}] validation passed")
            return result

        print(f"  [attempt {attempt + 1}] validation FAILED: {error}")
        if attempt == max_retries:
            print("  [giving up] max retries exhausted")
            return result

        # Retry-with-error-feedback: send back the assistant's own tool
        # call, then a tool_result containing the SPECIFIC error — not
        # a generic "please try again." The model can now re-read the
        # same source with a concrete thing to check.
        messages.append({"role": "assistant", "content": response.content})
        messages.append({
            "role": "user",
            "content": [{
                "type": "tool_result",
                "tool_use_id": tool_block.id,
                "content": f"Validation failed: {error}",
                "is_error": True,
            }],
        })

    return result


def demonstrate_retry_that_helps():
    print("=== Case A: format/structural error — retry CAN help ===\n")
    result = extract_with_retry(INVOICE_WITH_FIXABLE_ISSUE)
    print(f"\nFinal extraction: {result}\n")


def demonstrate_retry_that_cannot_help():
    print("=== Case B: info genuinely absent from source — retry CANNOT help ===\n")
    response = client.messages.create(
        model="claude-sonnet-5",
        max_tokens=1024,
        tools=[EXTRACT_INVOICE_TOOL],
        tool_choice={"type": "tool", "name": "extract_invoice"},
        messages=[{
            "role": "user",
            "content": f"Extract the invoice fields from this document:\n\n{INVOICE_WITHOUT_PO_NUMBER}"
        }],
    )
    tool_block = next(b for b in response.content if b.type == "tool_use")
    result = tool_block.input
    po = result.get("purchase_order_number")
    print(f"purchase_order_number extracted as: {po!r}")
    print(
        "Correct outcome is null/None here — the document never states a PO number.\n"
        "Retrying with 'the PO number field is empty, please try again' would NOT\n"
        "produce a real PO number; it would just pressure the model to fabricate one.\n"
        "The nullable field (Task 4.3) plus accepting null (Task 4.4) is the fix,\n"
        "not a retry loop."
    )


if __name__ == "__main__":
    demonstrate_retry_that_helps()
    print()
    demonstrate_retry_that_cannot_help()

File: 04_Prompt_Engineering_and_Structured_Output/Exercises/typescript/02_validation_retry_loop.ts
/**
 * Exercise 2 — Forced Tool Ordering + Validation/Retry (Tasks 4.3 & 4.4)
 * Domain: Prompt Engineering & Structured Output
 *
 * Demonstrates:
 *   - Forcing a SPECIFIC named tool ({"type": "tool", "name": "extract_metadata"})
 *     to guarantee a metadata-extraction step runs before a later
 *     enrichment step depends on its output
 *   - A self-correction schema design: extracting BOTH stated_total and
 *     calculated_total (computed independently) so a discrepancy is
 *     visible without a second model call
 *   - A conflict_detected boolean for genuinely inconsistent source data
 *   - retry-with-error-feedback for a format/structural mismatch
 *
 * Run: npx ts-node --esm 02_validation_retry_loop.ts
 */

import Anthropic from '@anthropic-ai/sdk';

const client = new Anthropic();

interface DocumentMetadata {
  document_type: 'invoice' | 'receipt' | 'purchase_order' | 'other';
  document_id: string | null;
}

interface InvoiceExtraction {
  vendor_name: string;
  line_items: { description: string; amount: number }[];
  stated_total: number;
  calculated_total: number; // model computes this itself from line_items
  conflict_detected: boolean; // true if stated_total != calculated_total
}

const METADATA_TOOL = {
  name: 'extract_metadata',
  description: 'Identify the basic document type and ID before deeper extraction.',
  input_schema: {
    type: 'object' as const,
    properties: {
      document_type: { type: 'string', enum: ['invoice', 'receipt', 'purchase_order', 'other'] },
      document_id: { type: ['string', 'null'] }
    },
    required: ['document_type']
  }
};

const INVOICE_TOOL = {
  name: 'extract_invoice',
  description: 'Extract invoice line items and totals, flagging any total mismatch.',
  input_schema: {
    type: 'object' as const,
    properties: {
      vendor_name: { type: 'string' },
      line_items: {
        type: 'array',
        items: {
          type: 'object',
          properties: { description: { type: 'string' }, amount: { type: 'number' } },
          required: ['description', 'amount']
        }
      },
      stated_total: { type: 'number', description: 'The total as printed on the document.' },
      calculated_total: { type: 'number', description: 'Sum of all line_items amounts, computed independently.' },
      conflict_detected: { type: 'boolean', description: 'true if stated_total and calculated_total differ.' }
    },
    required: ['vendor_name', 'line_items', 'stated_total', 'calculated_total', 'conflict_detected']
  }
};

// Deliberately inconsistent: the printed total ($500) does not match
// what the line items actually sum to ($480) — genuine source-data
// inconsistency, not an extraction mistake.
const INVOICE_TEXT = `
INVOICE #3390 — Marlowe Design Studio
  Logo design ......... $300.00
  Brand guidelines .... $180.00
TOTAL DUE: $500.00
`;

async function extractMetadataFirst(documentText: string): Promise<DocumentMetadata> {
  // Forced, specific tool: this step must run and must be THIS tool,
  // because a later enrichment step (routing to the right extraction
  // schema) depends on knowing document_type before anything else runs.
  const response = await client.messages.create({
    model: 'claude-sonnet-5',
    max_tokens: 256,
    tools: [METADATA_TOOL],
    tool_choice: { type: 'tool', name: 'extract_metadata' },
    messages: [{ role: 'user', content: `Identify the document type:\n\n${documentText}` }]
  });
  const block = response.content.find((b) => b.type === 'tool_use');
  if (!block || block.type !== 'tool_use') throw new Error('Expected tool_use');
  return block.input as DocumentMetadata;
}

async function extractInvoiceWithSelfCheck(documentText: string): Promise<InvoiceExtraction> {
  const response = await client.messages.create({
    model: 'claude-sonnet-5',
    max_tokens: 1024,
    tools: [INVOICE_TOOL],
    tool_choice: { type: 'tool', name: 'extract_invoice' },
    messages: [{ role: 'user', content: `Extract invoice fields, computing calculated_total independently from the line items:\n\n${documentText}` }]
  });
  const block = response.content.find((b) => b.type === 'tool_use');
  if (!block || block.type !== 'tool_use') throw new Error('Expected tool_use');
  return block.input as InvoiceExtraction;
}

async function main() {
  console.log('=== Step 1: forced extract_metadata runs before enrichment ===\n');
  const metadata = await extractMetadataFirst(INVOICE_TEXT);
  console.log(JSON.stringify(metadata, null, 2));

  if (metadata.document_type !== 'invoice') {
    console.log('Not an invoice — would route to a different extraction schema here.');
    return;
  }

  console.log('\n=== Step 2: invoice extraction with stated_total vs calculated_total self-check ===\n');
  const invoice = await extractInvoiceWithSelfCheck(INVOICE_TEXT);
  console.log(JSON.stringify(invoice, null, 2));

  if (invoice.conflict_detected) {
    console.log(
      `\nconflict_detected = true: stated_total (${invoice.stated_total}) != ` +
      `calculated_total (${invoice.calculated_total}). This is a SOURCE DATA ` +
      `inconsistency (the document itself is self-contradictory), not an ` +
      `extraction error — retrying will not resolve it. Route to human review ` +
      `with both values attached.`
    );
  } else {
    console.log('\nNo conflict — stated and calculated totals agree.');
  }
}

main();
